### DISCUSSION TITLE
Missing CVEs from bitnami database

### DISCUSSION THREAD

**Discussion by fmulero** [aquasecurity/trivy#6954] [Bugs]
### Description

Version 0.52.2 seems not being reporting CVEs declared in [bitnami/vulndb](https://github.com/bitnami/vulndb) database.

### Desired Behavior

As in versions 0.49.1, CVEs declared in [bitnami/vulndb](https://github.com/bitnami/vulndb) are reported by trivy.

### Actual Behavior

Some CVEs reported on [bitnami/vulndb](https://github.com/bitnami/vulndb) database are missing.

### Reproduction Steps

```bash
$ curl -sLo /tmp/solr-8.11.3-3-linux-amd64-debian-12.tar.gz https://downloads.bitnami.com/files/stacksmith/solr-8.11.3-3-linux-amd64-debian-12.tar.gz
$ tar -xzf /tmp/solr-8.11.3-3-linux-amd64-debian-12.tar.gz -C /tmp
$ trivy --version && trivy sbom /tmp/solr-8.11.3-linux-amd64-debian-12/files/solr/.spdx-solr.spdx --format json --debug > /tmp/vulns.0.52.json
Version: 0.52.2
Vulnerability DB:
  Version: 2
  UpdatedAt: 2024-06-18 06:11:42.309175838 +0000 UTC
  NextUpdate: 2024-06-18 12:11:42.309175167 +0000 UTC
  DownloadedAt: 2024-06-18 06:55:43.049837 +0000 UTC
Java DB:
  Version: 1
  UpdatedAt: 2024-05-27 01:07:06.626428609 +0000 UTC
  NextUpdate: 2024-05-30 01:07:06.626428419 +0000 UTC
  DownloadedAt: 2024-05-27 09:22:52.321385 +0000 UTC
2024-06-18T11:35:24+02:00       DEBUG   Parsed severities       severities=[UNKNOWN LOW MEDIUM HIGH CRITICAL]
2024-06-18T11:35:24+02:00       DEBUG   Ignore statuses statuses=[]
2024-06-18T11:35:24+02:00       DEBUG   Cache dir       dir="/Users/fmulero/Library/Caches/trivy"
2024-06-18T11:35:24+02:00       DEBUG   DB update was skipped because the local DB is the latest
2024-06-18T11:35:24+02:00       DEBUG   DB info schema=2 updated_at=2024-06-18T06:11:42.309175838Z next_update=2024-06-18T12:11:42.309175167Z downloaded_at=2024-06-18T06:55:43.049837Z
2024-06-18T11:35:24+02:00       INFO    Vulnerability scanning is enabled
2024-06-18T11:35:24+02:00       DEBUG   Vulnerability type      type=[os library]
2024-06-18T11:35:24+02:00       DEBUG   Enabling misconfiguration scanners      scanners=[]
2024-06-18T11:35:24+02:00       INFO    Detected SBOM format    format="spdx-json"
2024-06-18T11:35:24+02:00       DEBUG   OS is not detected.
2024-06-18T11:35:24+02:00       DEBUG   Detected OS: unknown
2024-06-18T11:35:24+02:00       INFO    Number of language-specific files       num=2
2024-06-18T11:35:24+02:00       INFO    [bitnami] Detecting vulnerabilities...
2024-06-18T11:35:24+02:00       DEBUG   [bitnami] Scanning packages for vulnerabilities file_path=""
2024-06-18T11:35:24+02:00       INFO    [jar] Detecting vulnerabilities...
2024-06-18T11:35:24+02:00       DEBUG   [jar] Scanning packages for vulnerabilities     file_path=""
$ /tmp/trivy --version && /tmp/trivy sbom /tmp/solr-8.11.3-linux-amd64-debian-12/files/solr/.spdx-solr.spdx --format json --debug > /tmp/vulns.0.49.json
Version: 0.49.1
Vulnerability DB:
  Version: 2
  UpdatedAt: 2024-06-18 06:11:42.309175838 +0000 UTC
  NextUpdate: 2024-06-18 12:11:42.309175167 +0000 UTC
  DownloadedAt: 2024-06-18 06:55:43.049837 +0000 UTC
Java DB:
  Version: 1
  UpdatedAt: 2024-05-27 01:07:06.626428609 +0000 UTC
  NextUpdate: 2024-05-30 01:07:06.626428419 +0000 UTC
  DownloadedAt: 2024-05-27 09:22:52.321385 +0000 UTC
2024-06-18T11:36:14.578+0200    DEBUG   Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2024-06-18T11:36:14.578+0200    DEBUG   Ignore statuses {"statuses": null}
2024-06-18T11:36:14.620+0200    DEBUG   cache dir:  /Users/fmulero/Library/Caches/trivy
2024-06-18T11:36:14.621+0200    DEBUG   DB update was skipped because the local DB is the latest
2024-06-18T11:36:14.621+0200    DEBUG   DB Schema: 2, UpdatedAt: 2024-06-18 06:11:42.309175838 +0000 UTC, NextUpdate: 2024-06-18 12:11:42.309175167 +0000 UTC, DownloadedAt: 2024-06-18 06:55:43.049837 +0000 UTC
2024-06-18T11:36:14.622+0200    INFO    Vulnerability scanning is enabled
2024-06-18T11:36:14.622+0200    DEBUG   Vulnerability type:  [os library]
2024-06-18T11:36:14.622+0200    DEBUG   Enabling misconfiguration scanners: []
2024-06-18T11:36:14.634+0200    INFO    Detected SBOM format: spdx-json
2024-06-18T11:36:14.735+0200    DEBUG   OS is not detected.
2024-06-18T11:36:14.735+0200    DEBUG   Detected OS: unknown
2024-06-18T11:36:14.735+0200    INFO    Number of language-specific files: 2
2024-06-18T11:36:14.735+0200    INFO    Detecting bitnami vulnerabilities...
2024-06-18T11:36:14.735+0200    DEBUG   Detecting library vulnerabilities, type: bitnami, path: 
2024-06-18T11:36:14.738+0200    INFO    Detecting jar vulnerabilities...
2024-06-18T11:36:14.738+0200    DEBUG   Detecting library vulnerabilities, type: jar, path:
$ diff -Nau /tmp/vulns.0.49.json /tmp/vulns.0.52.json
--- /tmp/vulns.0.49.json        2024-06-18 11:36:14
+++ /tmp/vulns.0.52.json        2024-06-18 11:35:24
@@ -1,6 +1,6 @@
 {
   "SchemaVersion": 2,
-  "CreatedAt": "2024-06-18T11:36:14.793408+02:00",
+  "CreatedAt": "2024-06-18T11:35:24.67086+02:00",
   "ArtifactName": "/tmp/solr-8.11.3-linux-amd64-debian-12/files/solr/.spdx-solr.spdx",
   "ArtifactType": "spdx",
   "Metadata": {
@@ -17,313 +17,18 @@
   },
   "Results": [
     {
-      "Target": "",
-      "Class": "lang-pkgs",
-      "Type": "bitnami",
-      "Vulnerabilities": [
-        {
-          "VulnerabilityID": "CVE-2023-44487",
-          "PkgName": "solr",
-          "PkgIdentifier": {
-            "PURL": "pkg:bitnami/solr@8.11.3-3?arch=amd64\u0026distro=debian-12"
-          },
-          "InstalledVersion": "8.11.3-3",
-          "FixedVersion": "9.4.0",
-          "Status": "fixed",
-          "Layer": {},
-          "SeveritySource": "bitnami",
-          "PrimaryURL": "https://avd.aquasec.com/nvd/cve-2023-44487",
-          "DataSource": {
-            "ID": "bitnami",
-            "Name": "Bitnami Vulnerability Database",
-            "URL": "https://github.com/bitnami/vulndb"
-          },
-          "Title": "HTTP/2: Multiple HTTP/2 enabled web servers are vulnerable to a DDoS attack (Rapid Reset Attack)",
-          "Description": "The HTTP/2 protocol allows a denial of service (server resource consumption) because request cancellation can reset many streams quickly, as exploited in the wild in August through October 2023.",
-          "Severity": "HIGH",
-          "CweIDs": [
-            "CWE-400"
-          ],
-          "VendorSeverity": {
-            "alma": 3,
-            "amazon": 3,
-            "bitnami": 3,
-            "cbl-mariner": 3,
-            "ghsa": 2,
-            "nvd": 3,
-            "oracle-oval": 3,
-            "photon": 3,
...
```

### Target

SBOM

### Scanner

Vulnerability

### Output Format

JSON

### Mode

Standalone

### Debug Output

```bash
Please see reproduction steps
```

### Operating System

macOS Sonoma

### Version

```bash
Version: 0.52.2
Vulnerability DB:
  Version: 2
  UpdatedAt: 2024-06-18 06:11:42.309175838 +0000 UTC
  NextUpdate: 2024-06-18 12:11:42.309175167 +0000 UTC
  DownloadedAt: 2024-06-18 06:55:43.049837 +0000 UTC
Java DB:
  Version: 1
  UpdatedAt: 2024-05-27 01:07:06.626428609 +0000 UTC
  NextUpdate: 2024-05-30 01:07:06.626428419 +0000 UTC
  DownloadedAt: 2024-05-27 09:22:52.321385 +0000 UTC
```

### Checklist

- [ ] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by DmitriyLewen** [comment]
Hello @fmulero 
Thanks for your report!

We use `package.name` field to get package name (to avoid case difference issues):
https://github.com/aquasecurity/trivy/blob/c3192f061d7e84eaf38df8df7c879dc00b4ca137/pkg/sbom/io/decode.go#L249-L266

In you case package name is `Apache Solr`:
```bash
➜  cat /tmp/solr-8.11.3-linux-amd64-debian-12/files/solr/.spdx-solr.spdx | grep '"SPDXID": "SPDXRef-solr",' -A 2
    "SPDXID": "SPDXRef-solr",
    "spdxVersion": "SPDX-2.3",
    "creationInfo": {
--
            "SPDXID": "SPDXRef-solr",
            "name": "Apache Solr",
            "versionInfo": "8.11.3-3",
```

But [bitnami/vulndb](https://github.com/bitnami/vulndb) database uses `solr` name:
https://github.com/bitnami/vulndb/blob/70557cdbd4584f16849f529c7fdafe6d81e6309f/data/solr/BIT-solr-2023-44487.json

This is why Trivy doesn't detect this vulnerability.

Regards, Dmitriy

**Reply by fmulero** [reply]
Thanks a lot @DmitriyLewen for your quick response. Then, the PURL is not enough to identify the package, names should match also. Is there any other field to take into consideration? 

If we align the names in the Bitnami side, changing all those names for all the assets in the bitnami/vulndb, Do you see any inconvenience from your side?

**Reply by DmitriyLewen** [reply]
If you have options to change the Bitnami product, I think the best way is to fix the `package.name` field in the spdx file.
Because this is a strange case where `package.name` and the package name from `purl` are different.

**Reply by fmulero** [reply]
We agree that the name in bitnami/vulndb and the SPDX files should match but changing the `package.name` in the spdx file will work only for new releases, existing packages will continue failing. If we change only the SPDX file, I think we should consider do changes in Trivy side to avoid problems with existing releases (AFAIK the PURL should be enough to identify the package).

**Reply by DmitriyLewen** [reply]
By agreement with `bitnami`, perhaps we can only use package names from purl for `bitnami` packages.
@knqyf263 wdyt?

**Reply by knqyf263** [reply]
We need a package name as PURL is case-insensitive. It sounds reasonable only for Bitnami packages as a workaround.

**Reply by DmitriyLewen** [reply]
It looks like the best way for old spdx files is to create a utility to update the `package.name` field.

**Reply by knqyf263** [reply]
>It looks like the best way for old spdx files is to create a utility to update the package.name field.

While I think Bitnami needs to fix it, I'm fine with working around the problem for old SPDX files on the Trivy side.

**Reply by DmitriyLewen** [reply]
@fmulero is there a way to understand where "old" and where "new" spdx files are?
Or do you think we need to use purl for all sdpx files?

**Reply by fmulero** [reply]
I've just opened a internal discussion, I'll come back with any news. 

Just for curiosity, how does version 0.49.1 work? Is this version using the purl instead of the name?

**Reply by DmitriyLewen** [reply]
Right. we missed the package name for `spdx`.
But we fixed it after adding `bom.Core` (https://github.com/aquasecurity/trivy/pull/6240 + https://github.com/aquasecurity/trivy/pull/6310)

**Reply by fmulero** [reply]
We are going to do the changes in the SPDX files to use the same `package.name`s set in the `bitnami/vulndb` database, but we can not change the spdx files for all the assets already released (we could force new releases but we can not change the existing ones).

The `purl` field will be kept and it is aligned with the content of the `bitnami/vulndb`,  so you can trust on this field to match the SPDX packages with the Bitnami OSV.  I think this fallback could be dropped (if you don't want to keep it) in 6-9 months, to ensure that all packages have several releases with the new format.

**Reply by DmitriyLewen** [reply]
@knqyf263 are you agree with using `purl` to get  package name **for Bitnami packages only**?
If yes, I will create PR for these changes.

**Reply by knqyf263** [reply]
I'm fine.

**Comment by DmitriyLewen** [comment]
Track #6981
